[AutoDiff] Support differentiation of conditionals.

dan-zheng opened the pull request:
- Support control flow in adjoint generation.
  - Make adjoint value/buffer mappings be per basic block.
  - Change `AdjointValue` to not be move-only. Original values from
    different basic blocks may share the same `AdjointValue`.
  - Propagate adjoint values from active bb arguments to predecessor
    terminator operands.
  - Propagate adjoint values/buffers of dominated active values/buffers
    to predecessor blocks.
    - For active values: propagate adjoint values as adjoint bb arguments.
    - For active buffers: propagate adjoint buffers via `copy_addr`.
- Revamp `AdjointEmitter` handling of `begin_access` and `end_access`.
  - `getAdjointValue` of `begin_access` now returns the adjoint base buffer.
    Previously, it returned a `begin_access` of the adjoint base buffer
    without generating a corresponding `end_access`.
  - `AdjointEmitter::visitBeginAccessInst` now generates no code.
  - `AdjointEmitter::visitEndAccessInst` now does nothing.
- Add various control flow differentiation tests.
  - Test differentiation of conditionals (nested), recursion,
    `var` allocations (tuples, structs).
  - Add negative leakchecking tests.

Todos:
- Fix adjoint value/buffer propagation memory leaks.
- Add more tests (adjoint SIL, leakchecking).
- Support differentiation of enum-related instructions and loops.

dan-zheng committed:
Support differentiation of conditionals.

- Support control flow in adjoint generation.
  - Make adjoint value mapping be per basic block.
  - Change `AdjointValue` to not be move-only. Original values from
    different basic blocks can share the same `AdjointValue`.
  - Propagate adjoint values from active basic block arguments to
    predecessor terminator operands.
  - Propagate adjoint values of active values to adjoint successor blocks.
- Add runtime control flow differentiation tests.

dan-zheng commented:
This is a work in-progress; there are at least two bugs.
* Generic adjoint functions with control flow fail verification. This should be a simple type-remapping problem.
  * grep `TODO: Fix generics crash (related to type remapping) in AdjointEmitter` in tests.
* Gradient values are sometimes incorrectly 0.
  * grep `FIXME: Debug incorrect "0" gradient values wrt x` in tests.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3875,25 +3904,36 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+  /// corresponding adjoint forwarding basic blocks. Forwarding basic blocks
+  /// take additional arguments in addition to the predecessor enum argument.
+  DenseMap<std::pair<SILBasicBlock *, SILBasicBlock *>, SILBasicBlock *>
+      adjointForwardingBBMap;
```
Rename to `adjointTrampolineBBMap`?
"Trampoline" is standard terminology for "basic blocks that just branch" - see `SILBasicBlock::isTrampoline()`.
Though adjoint trampoline BBs may also materialize `AdjointValue`s to `SILValue`s before branching.

rxwei replied:
Yes

dan-zheng replied:
Done.

dan-zheng committed:
Fix type remapping crash, expose "use of un-dominated operand" crash.

Expose new crash in adjoint verification:
```
SIL verification failed: instruction isn't dominated by its operand:
properlyDominates(valueI, I)
```

dan-zheng committed:
Rename "forwarding basic block" to "trampoline basic block".

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1754,6 +1766,30 @@ Activity DifferentiableActivityInfo::getActivity(
+    if (isActive(arg, indices))
+      addActiveValue(arg);
+  for (auto &inst : *bb) {
+    for (auto op : inst.getOperandValues())
```
If you iterate over each instruction's operands, then it's not just collecting active values that are defined in the basic block, but also all active values that are used in the BB. Is that the intended behavior?

dan-zheng replied:
Yes, that is intended behavior.

If `%0` is an argument of `bb0` but is used in `bb1`, then `bb1_adj` (adjoint block of bb1) needs `%0_adj` as a BB argument.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3875,25 +3904,36 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-  DenseMap<SILValue, AdjointValue> valueMap;
+  /// Mapping from original basic blocks to a mapping from original values to
+  /// their corresponding adjoint values.
+  DenseMap<SILBasicBlock *, DenseMap<SILValue, AdjointValue>> valueMap;
```
Unless you need to index into this map using a BB and iterate over the nested map's keys and values, I'd suggest using a tuple (`std::pair<SILBasicBlock *, SILValue>`) as the key.

Plus, how is the "original basic block" and "original value" related? Is the value expected to be defined in the BB?

dan-zheng replied:
Using `std::pair<SILBasicBlock *, SILValue>` as the key seems fine to me.

> Plus, how is the "original basic block" and "original value" related? Is the value expected to be defined in the BB?

The original value is expected to be dominated (not necessarily used) in the original basic block. For example, there may exist a key `(bb3, %0)` even though `%0` is defined in `bb0`.

dan-zheng replied:
Conclusion: `DenseMap<std::pair<SILBasicBlock *, SILValue>, AdjointValue>` is the new key type. This is more efficient than `DenseMap<SILBasicBlock *, DenseMap<SILValue, AdjointValue>>`.

rxwei commented:
FYI: Since @dan-zheng is OOO, I'm taking over this today to make progress.

rxwei committed:
Address comments.

rxwei committed:
Merge branch 'tensorflow' of github.com:apple/swift into autodiff-control-flow

rxwei committed:
Merge branch 'tensorflow' of github.com:apple/swift into autodiff-control-flow

rxwei committed:
Merge branch 'tensorflow' of github.com:apple/swift into autodiff-control-flow

rxwei committed:
Expose incorrect active buffer handling.

Also fix `getRemappedTangentType` to handle addresses.

dan-zheng committed:
Fix active adjoint buffer handling.

- Make adjoint buffer mapping per basic block.
- For active buffers in original blocks, allocate and zero initialize adjoint
  buffers in adjoint entry.
- Propagate adjoint buffers via `copy_addr` before branching in adjoint blocks.
- Re-enable fixed tests.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4380,11 +4386,17 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-#error "FIXME: create local allocations"
+          // Allocate and zero initialize a new local buffer using
+          // `getAdjointBuffer`.
+          builder.setInsertionPoint(adjoint.getEntryBlock());
```
`builder.setInsertionPoint(adjoint.getEntryBlock());` is crucial because `getAdjointBuffer` currently assumes that `builder` is at a valid insertion point before `emitZeroIndirect`.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4234,22 +4239,23 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     // allocation, or at the start of the adjoint function's entry if no local
     // allocations exist yet.
     localAllocBuilder.setInsertionPoint(
+        getAdjoint().getEntryBlock(),
```
This line is crucial to prevent an `Assertion failed: (!NodePtr->isKnownSentinel())` crash.

dan-zheng committed:
Change adjoint function local allocation insertion order.

Insert local allocations before previous local allocations to ensure
that allocation and zero initialization instructions are grouped together.

Fixes test/AutoDiff/generics.swift.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4216,11 +4216,15 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   }
 
   SILBasicBlock::iterator getNextFunctionLocalAllocationInsertionPoint() {
```
Might be better to turn this into a function that allocates a local buffer directly: ‘createFunctionLocalBuffer(SILType)’.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4418,23 +4516,95 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+          SmallVector<SILValue, 8> arguments;
+          arguments.push_back(adjointSuccBB->getArguments().front());
+          // Propagate adjoint values/buffers of active values/buffers.
+          for (unsigned i : indices(activeValues[predBB])) {
```
Note: this adjoint value/buffer propagation logic should be hoisted out of this `if-else` block so that it executes regardless of whether there's a trampoline block.

dan-zheng replied:
Furthermore, it's necessary to propagate more than just the "adjoint values of the active values in `predBB`".
Consider the following function:
```swift
@differentiable
@_silgen_name("cond_empty")
func cond_empty(_ x: Float) -> Float {
  if x > 0 {
    // Generates empty basic blocks between entry/exit.
  }
  return x
}
```

```
// cond_empty
sil hidden [differentiable source 0 wrt 0 jvp @AD__cond_empty__jvp_src_0_wrt_0 vjp @AD__cond_empty__vjp_src_0_wrt_0] @cond_empty : $@convention(thin) (Float) -> Float {
// %0                                             // users: %5, %20, %1
bb0(%0 : $Float):
  debug_value %0 : $Float, let, name "x", argno 1 // id: %1
  %2 = metatype $@thin Float.Type
  %3 = metatype $@thick Float.Type                // user: %13
  %4 = alloc_stack $Float                         // users: %5, %15, %13
  store %0 to %4 : $*Float                        // id: %5
  %6 = integer_literal $Builtin.IntLiteral, 0     // user: %9
  %7 = metatype $@thin Float.Type                 // user: %9
  // function_ref Float.init(_builtinIntegerLiteral:)
  %8 = function_ref @$sSf22_builtinIntegerLiteralSfBI_tcfC : $@convention(method) (Builtin.IntLiteral, @thin Float.Type) -> Float // user: %9
  %9 = apply %8(%6, %7) : $@convention(method) (Builtin.IntLiteral, @thin Float.Type) -> Float // user: %11
  %10 = alloc_stack $Float                        // users: %11, %14, %13
  store %9 to %10 : $*Float                       // id: %11
  // function_ref static FloatingPoint.> infix(_:_:)
  %12 = function_ref @$sSFsE1goiySbx_xtFZ : $@convention(method) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> Bool // user: %13
  %13 = apply %12<Float>(%4, %10, %3) : $@convention(method) <τ_0_0 where τ_0_0 : FloatingPoint> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, @thick τ_0_0.Type) -> Bool // user: %16
  dealloc_stack %10 : $*Float                     // id: %14
  dealloc_stack %4 : $*Float                      // id: %15
  %16 = struct_extract %13 : $Bool, #Bool._value  // user: %17
  cond_br %16, bb1, bb2                           // id: %17

bb1:                                              // Preds: bb0
  br bb3                                          // id: %18

bb2:                                              // Preds: bb0
  br bb3                                          // id: %19

bb3:                                              // Preds: bb1 bb2
  return %0 : $Float                              // id: %20
} // end sil function 'cond_empty'
```

Since `%0` is active in `bb0` and `bb3`, `%0_adj` must be somehow be propagated from `bb3_adj` to `bb0_adj` - perhaps via the adjoint value map for `bb1` and `bb2`.

dan-zheng replied:
Resolved in https://github.com/apple/swift/pull/25057/commits/bd8d3190290840d0d05b6f32ee804072beab305f.
The solution is to track "dominated active values", not just "used active values" in original basic blocks.

dan-zheng committed:
Minor edit.

dan-zheng committed:
Fix/simplify adjoint value propagation logic.

- Record *dominated* active values per basic block, not just
  active values used in basic blocks.
- Remove `visitAllocStack` and `visitDeallocStack` from `AdjointEmitter`.
  - These fall out from active buffer handling.
- Add nested conditional and recursion tests.
- Remove "-differentiable-enable-control-flow" flag, now that
  limited control flow support seems robust.

dan-zheng committed:
Merge remote-tracking branch 'apple/tensorflow' into autodiff-control-flow

dan-zheng commented:
Ready for re-review. Adjoint value/buffer propagation logic has been fixed and simplified: all test cases pass.

The core control flow differentiation algorithm can be summarized as: **create adjoint bb arguments (for values) and "join buffers" (for buffers) to propagate adjoints of dominated active values/buffers**.

rxwei commented on test/AutoDiff/control_flow.swift:
```diff
@@ -1,79 +1,138 @@
-@_silgen_name("cond")
-func cond(_ x: Float) -> Float {
-  if x > 0 {
+var ControlFlowTests = TestSuite("ControlFlow")
```
`guard` tests should be added.

dan-zheng replied:
Added in https://github.com/apple/swift/pull/25057/commits/bde639fd10de0912da43c49dfcf7417d279da4ff.

dan-zheng committed:
Add guard statement tests.

dan-zheng commented on test/AutoDiff/control_flow.swift:
```diff
@@ -39,6 +39,42 @@ ControlFlowTests.test("Conditionals") {
+    return y * y
+  }
+  expectEqual((0, 10), gradient(at: 4, 5, in: guard2))
+  expectCrash {
```
Note: `expectCrash` doesn't seem to work as intended: it doesn't produce a failure even when the trailing closure doesn't crash.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Minor edits.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4077,35 +4118,117 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+    // arguments.
+    auto *domAnalysis =
+        getContext().getPassManager().getAnalysis<DominanceAnalysis>();
+    auto *domInfo = domAnalysis->get(&original);
```
Todo: make `domInfo` a member of `AdjointEmitter`, just like `postOrderInfo`.
Will commit after CI passes.

dan-zheng commented on test/AutoDiff/autodiff_diagnostics.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-swift-frontend -emit-sil -verify %s
+// RUN: %target-swift-frontend -emit-sil -verify -Xllvm -differentiation-enable-control-flow %s
```
Todo: revert the accidental RUN line change. `-Xllvm -differentiation-enable-control-flow` should go away soon.
Will commit after CI passes.

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4193,23 +4317,98 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+              // Propagate adjoint buffers using `copy_addr`.
+              auto adjBuf = getAdjointBuffer(bb, activeValue);
+              auto succAdjBuf = getAdjointBuffer(predBB, activeValue);
+              builder.createCopyAddr(
```
We should put more consideration into cleanup handling. The current implementation may cause segfaults. When you copy something to a function-local buffer, you should _transfer_ the ownership of the cleanup to the destination buffer so that the underlying value only gets cleaned up when the buffer is getting deallocated.

dan-zheng replied:
I see. Is this the correct logic or no?
```cpp
              auto adjBuf = getAdjointBuffer(bb, activeValue);
              auto succAdjBuf = getAdjointBuffer(predBB, activeValue);
              succAdjBuf.setCleanup(makeCleanupFromChildren(
                  {adjBuf.getCleanup(), succAdjBuf.getCleanup()}));
```

rxwei replied:
To step back, there are two problems:
1. How do you guarantee `activeValue` has never been an argument to `getAdjointBuffer(predBB, activeValue)` before? That is, will there be a case where `getAdjointBuffer(predBB, activeValue)` is called on the same `activeValue` twice? If that does happen, this code will simply rewrite the the existing adjoint value, which may give you a miscalculation error. Otherwise, it's worth thinking about how the code can be structured in a way that the reader understands this is the first time such a buffer is created for this particular adjoint value.
2. Back to cleanups, the way you should deal with cleanups is directly related to (1). If there are multiple calls to `getAdjointBuffer(predBB, activeValue)` for the same `activeValue`, then you should probably do an `AdditiveArithmetic.+=(&succAdjBuf, adjBuf)` and transfer no cleanup. Otherwise, do `succAdjBuf.setCleanup(adjBuf.getCleanup())` and set `adjBuf`'s cleanup to `nullptr`.

In any case, I think (1) should be resolved first.

dan-zheng replied:
We peer programmed and discussed the two problems in depth.

We found that `getAdjointBuffer(predBB, activeValue)` is called multiple times for the same `activeValue`. But performing `AdditiveArithmetic.+=(&predAdjBuf, adjBuf)` instead of `copy_addr(adjBuf, predAdjBuf)` lead to computation errors.

Progress in https://github.com/apple/swift/pull/25057/commits/49937d93f87989f9d46d29a956bb42923ca143d2: `var` + control flow tests are fixed, but cleanups during adjoint value/buffer propagation are not handled, leading to memory leaks.

Negative leak checking tests added to `test/AutoDiff/leakchecking.swift`. @rxwei offered to help fix the cleanup logic in a follow-up PR.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4623,13 +4807,14 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   ///   ...
   void visitTupleInst(TupleInst *ti) {
-    auto av = takeAdjointValue(ti);
+    auto *bb = ti->SILInstruction::getParentBlock();
```
Xcode completions aren't always nice.
```suggestion
    auto *bb = ti->getParentBlock();
```

dan-zheng replied:
This was done to avoid a compilation error, not an Xcode completion. Do you have a better suggestion?
```
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:4818:20: error: non-static member 'getParentBlock' found in multiple base-class subobjects of type 'swift::SILNode':
    class swift::TupleInst -> InstructionBaseWithTrailingOperands<enum SILInstructionKind::TupleInst, class swift::TupleInst, class swift::OwnershipForwardingSingleValueInst> -> InstructionBase<(enum swift::SILInstructionKind)84, class swift::OwnershipForwardingSingleValueInst> -> class swift::OwnershipForwardingSingleValueInst -> class swift::SingleValueInstruction -> class swift::SILInstruction -> class swift::SILNode
    class swift::TupleInst -> InstructionBaseWithTrailingOperands<enum SILInstructionKind::TupleInst, class swift::TupleInst, class swift::OwnershipForwardingSingleValueInst> -> InstructionBase<(enum swift::SILInstructionKind)84, class swift::OwnershipForwardingSingleValueInst> -> class swift::OwnershipForwardingSingleValueInst -> class swift::SingleValueInstruction -> class swift::ValueBase -> class swift::SILNode
    auto *bb = ti->getParentBlock();
                   ^
/Users/danielzheng/swift-dev/swift/include/swift/SIL/SILNode.h:429:18: note: member found by ambiguous name lookup
  SILBasicBlock *getParentBlock() const;
                 ^
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:4836:30: error: 'TupleType' does not refer to a value
        if (val.getType().is<TupleType>())
                             ^
/Users/danielzheng/swift-dev/swift/include/swift/AST/Module.h:76:9: note: declared here
  class TupleType;
        ^
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:4836:41: error: expected expression
        if (val.getType().is<TupleType>())
                                        ^
```

rxwei replied:
There's a method called `getParent()` also. Have you tried that one?

dan-zheng replied:
Oops, `getParent()` does work. I'll use that, thanks.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4663,18 +4848,19 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   ///                         |--- n-th element
   ///   adj[x] += tuple (0, 0, ..., adj[y], ..., 0, 0)
   void visitTupleExtractInst(TupleExtractInst *tei) {
+    auto *bb = tei->SILInstruction::getParentBlock();
```
```suggestion
    auto *bb = tei->getParentBlock();
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4692,50 +4878,22 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   //    Adjoint: adj[x] += adj[y]
   void visitLoadInst(LoadInst *li) {
-    auto adjVal = materializeAdjointDirect(takeAdjointValue(li), li->getLoc());
+    auto *bb = li->SILInstruction::getParentBlock();
```
```suggestion
    auto *bb = li->getParentBlock();
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4854,8 +5014,9 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     }
-    auto accessBuf = getAdjointBuffer(bai);
-    auto &sourceBuf = getAdjointBuffer(bai->getSource());
+    auto *bb = bai->SILInstruction::getParentBlock();
```
```suggestion
    auto *bb = bai->getParentBlock();
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4077,35 +4123,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+        activeBBValues.append(activeDomBBValues.begin(),
+                              activeDomBBValues.end());
+      }
+      SmallPtrSet<SILValue, 8> registered(activeBBValues.begin(),
```
How about a more common name: "visited"?

```suggestion
      SmallPtrSet<SILValue, 8> visited(activeBBValues.begin(),
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4801,7 +4960,8 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   //    Adjoint: adj[x] += adj[y]; adj[y] = 0
   void visitCopyAddrInst(CopyAddrInst *cai) {
-    auto &adjDest = getAdjointBuffer(cai->getDest());
+    auto *bb = cai->SILInstruction::getParentBlock();
```
```suggestion
    auto *bb = cai->getParentBlock();
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4768,7 +4926,8 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   //    Adjoint: adj[x] += load adj[y]; adj[y] = 0
   void visitStoreInst(StoreInst *si) {
-    auto &adjBuf = getAdjointBuffer(si->getDest());
+    auto *bb = si->SILInstruction::getParentBlock();
```
```suggestion
    auto *bb = si->getParentBlock();
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4867,7 +5028,8 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   //    Adjoint: adj[y] = begin_access inverse(access_kind) adj[x]
   void visitEndAccessInst(EndAccessInst *eai) {
-    auto adjBuf = getAdjointBuffer(eai->getSource());
+    auto *bb = eai->SILInstruction::getParentBlock();
```
```suggestion
    auto *bb = eai->getParentBlock();
```

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4881,14 +5043,15 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
   void visit##INST##Inst(INST##Inst *inst) { \
-    auto &adjBase = getAdjointBuffer(inst->getOperand()); \
-    auto &adjProj = getAdjointBuffer(inst); \
+    auto *bb = inst->SILInstruction::getParentBlock(); \
```
```suggestion
    auto *bb = inst->getParentBlock(); \
```

rxwei commented on test/AutoDiff/control_flow.swift:
```diff
@@ -1,79 +1,174 @@
+      }
+    }
+    return -y
+  }
```
Please add a test that manifests definitive initialization. That is, define a `let` with no initialization, and initialize it in all control flow paths. Also try `var` with some mutation in a branch.

dan-zheng replied:
`func cond2` above is a definitive initialization test - I think one such test is sufficient.
I'll add `var` tests soon.

dan-zheng replied:
`var` tests added in https://github.com/apple/swift/pull/25057/commits/806b4643a3c208275d4a3da8bcde6e7544fbd57b, fixed in https://github.com/apple/swift/pull/25057/commits/49937d93f87989f9d46d29a956bb42923ca143d2.

rxwei reviewed (commented):
I couldn't find any tests that are using `var`s. I think mutation should be tested in order for this PR to be ready.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4193,23 +4317,98 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+          // Propagate pullback struct argument.
+          trampolineArguments.push_back(adjointSuccBB->getArguments().front());
+          // Propagate adjoint values/buffers of active values/buffers.
+          for (unsigned i : indices(activeValues[predBB])) {
```
Thanks for the catch! Explicit indices are no longer necessary.

dan-zheng replied:
Also, renaming `active(XX)BBValues` to `(XX)bbActiveValues` everywhere, so that `ActiveValues` is one contiguous phrase.

dan-zheng committed:
Address review feedback.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -4193,23 +4317,98 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+            } else {
+              // Propagate adjoint buffers using `copy_addr`.
+              auto adjBuf = getAdjointBuffer(bb, activeValue);
+              auto succAdjBuf = getAdjointBuffer(predBB, activeValue);
```
The argument is `predBB`. Shouldn't the variable be called `predAdjBuf`?

dan-zheng replied:
I think `predAdjBuf` may be a better name, but both names are somewhat confusing.

I originally named the variable `succAdjBuf` because it's a buffer in the *adjoint successor block*. But naming it as an *adjoint buffer of the predecessor block* seems less confusing since `predBB` is an argument, as you mentioned.

I guess the code here could be described as:
1. Propagating adjoint values/buffers to predecessor blocks.
2. Propagating adjoint values/buffers to adjoint successor blocks.

1 seems more natural to me, so I'll change to `predAdjBuf`.

dan-zheng committed:
Add control flow `var` tests.

`var` tests currently crashing due to `AdjointEmitter` handling of
`begin_access`/`end_access`.

dan-zheng committed:
Fix `var` + control flow differentiation.

- Change `AdjointEmitter` handling of `begin_access` and `end_access`.
  - Change `AdjointEmitter::getAdjointValue` of `begin_access` instruction
    to return adjoint buffer of the base buffer. Previously, it returned a
    `begin_access` of the adjoint base buffer, without generating a
    corresponding `end_access` instruction, leading to a hole.
  - Change `AdjointEmitter::visitBeginAccessInst` to generate no code,
    only performing verification and propagating cleanups.
  - Change `AdjointEmitter::visitEndAccessInst` to do nothing.
- Fix and add `var` tests.
- Todo: fix control-flow related memory leaks (FIXMEs) related to adjoint
  value/buffer propagation.

dan-zheng committed:
Address review feedback.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented on test/AutoDiff/leakchecking.swift:
```diff
@@ -8,6 +8,13 @@ import DifferentiationUnittest
 
 var LeakCheckingTests = TestSuite("LeakChecking")
 
+/// Execute body, check expect leak count, and reset global leak count.
```
```suggestion
/// Execute body, check expected leak count, and reset global leak count.
```
Maybe fix in follow-up PR to unblock progress.

dan-zheng replied:
Fixed in https://github.com/apple/swift/pull/25057/commits/ca36c09d80d1db06495e3d68feea790c8db9ac96.

dan-zheng commented:
> To step back, there are two problems:
> 
> 1. How do you guarantee `activeValue` has never been an argument to `getAdjointBuffer(predBB, activeValue)` before? That is, will there be a case where `getAdjointBuffer(predBB, activeValue)` is called on the same `activeValue` twice? If that does happen, this code will simply rewrite the the existing adjoint value, which may give you a miscalculation error. Otherwise, it's worth thinking about how the code can be structured in a way that the reader understands this is the first time such a buffer is created for this particular adjoint value.
> 2. Back to cleanups, the way you should deal with cleanups is directly related to (1). If there are multiple calls to `getAdjointBuffer(predBB, activeValue)` for the same `activeValue`, then you should probably do an `AdditiveArithmetic.+=(&succAdjBuf, adjBuf)` and transfer no cleanup. Otherwise, do `succAdjBuf.setCleanup(adjBuf.getCleanup())` and set `adjBuf`'s cleanup to `nullptr`.
> 
> In any case, I think (1) should be resolved first.

We peer programmed and discussed the two problems in depth.

We found that `getAdjointBuffer(predBB, activeValue)` is called multiple times for the same `activeValue`. But performing `AdditiveArithmetic.+=(&predAdjBuf, adjBuf)` instead of `copy_addr(adjBuf, predAdjBuf)` lead to computation errors.

Progress in https://github.com/apple/swift/pull/25057/commits/49937d93f87989f9d46d29a956bb42923ca143d2: `var` + control flow tests are fixed, but cleanups during adjoint value/buffer propagation are not handled, leading to memory leaks.

Negative leak checking tests added to `test/AutoDiff/leakchecking.swift`. @rxwei offered to help fix the cleanup logic in a follow-up PR.

dan-zheng commented on test/AutoDiff/control_flow.swift:
```diff
@@ -1,79 +1,277 @@
+  expectEqual(2, gradient(at: -10, in: cond2))
+  expectEqual(0, gradient(at: -1337, in: cond2))
+
+  func cond_tuple_var(_ x: Float) -> Float {
```
Todo: add `struct_extract` and `struct_element_addr` tests.
I'll start writing tests - may land in follow-up PR to unblock progress.

dan-zheng replied:
Added in https://github.com/apple/swift/pull/25057/commits/9af6170f2cdd94813e272298d8845cdbdd7e7e47.

dan-zheng commented on test/AutoDiff/control_flow.swift:
```diff
@@ -1,79 +1,277 @@
+    let y: (Float, Float) = (x, x)
+    return y.0 + y.1
+  }
+  expectEqual(8, gradient(at: 4, in: cond2))
```
Oops, I messed up the function name here, and for `cond_tuple_var` below.
```suggestion
  expectEqual(8, gradient(at: 4, in: cond_tuple))
```
Will fix and retrigger CI.

rxwei reviewed (approved):
LGTM for merging as-is after existing issues are addressed. I’ll work on a fix for memory leaks.

dan-zheng committed:
Fix tests.

- Make sure `gradient` calls test the right function.
- Add struct `let` and `var` tests.

Note: tuple/struct tests are designed to compute `x + x`
(in convoluted ways) for easy verification.

Corresponding tuple/struct tests have essentially the same body
(e.g. `cond_nestedtuple_var` ~= `cond_nestedstruct_var`).

dan-zheng commented on test/AutoDiff/leakchecking.swift:
```diff
@@ -22,7 +29,58 @@ LeakCheckingTests.test("BasicVarLeakChecking") {
+}
+
+LeakCheckingTests.test("ControlFlow") {
+  testWithLeakChecking(expectedLeakCount: 0) {
```
This `cond_tuple_var` test is not effective because it operates on `Float` instead of `Tracked<Float>`.
Fixing it requires significant changes to the `Tracked` type. I'll fix and re-add it (and more tests) in a follow-up.

The control flow leakchecking tests below should be sufficient for testing memory leaks for now.

dan-zheng committed:
Temporarily remove non-functional `var` + control flow leakchecking test.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Add `Tensor` tests with `var` and control flow.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei commented:
@swift-ci please test tensorflow
